server/src/main/java/org/opensearch/plugins/ResourceAccessControlPlugin.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.accesscontrol.resources.RecipientType;
+import org.opensearch.accesscontrol.resources.Resource;
+import org.opensearch.accesscontrol.resources.ResourceSharing;
+import org.opensearch.accesscontrol.resources.ShareWith;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This plugin allows to control access to resources. It is used by the ResourcePlugins to check whether a user has access to a resource defined by that plugin.
+ * It also defines java APIs to list, share or revoke resources with other users.
+ * User information will be fetched from the ThreadContext.
+ *
+ * @opensearch.experimental
+ */
+public interface ResourceAccessControlPlugin {
+
+    /**
+     * Returns all accessible resources for current user for a given plugin index.
+     * @param resourceIndex index where the resource exists
+     * @param clazz class of the resource. Required to parse the resource object retrieved from resourceIndex. Must be a type of {@link Resource}
+     * @return set of {@link ResourceSharing} items accessible by current user.
+     */
+    default <T extends Resource> Set<T> getAccessibleResourcesForCurrentUser(String resourceIndex, Class<T> clazz) {
+        return Set.of();
+    }
+
+    /**
+     * Checks whether current user has permission to given resource.
+     *
+     * @param resourceId the resource on which access is to be checked
+     * @param resourceIndex where the resource exists
+     * @param scope the scope being requested
+     * @return true if current user has access, false otherwise
+     */
+    default boolean hasPermission(String resourceId, String resourceIndex, String scope) {
+        return true;
+    }
+
+    /**
```

Comment:
wdyt about implementing this and the method below as a REST API in the plugin that implements resource access control?

i.e. Security plugin could have an API

```
PUT /resource/{resource_type}/{id}/share_with
{
    "resource_id": "...",
    "resource_type": "...", // resource_type is 1 <-> 1 with resource index, but is a human readable type
    "share_with": { 
         "users": ["userB"]
    }
}
```

Does this need to be part of the interface?